### DISCUSSION TITLE
linux x86 交叉编译 arm64 二进制

### DISCUSSION THREAD

**Discussion by TOMO-CAT** [xmake-io/xmake#4985] [Q&A]
由于 clang 本身支持设置 target，只是缺少对应架构的 include 和 lib 文件夹，我先下载了对应的 sdk：
```bash
apt-get install gcc-arm-linux-gnueabi
```
![image](https://github.com/xmake-io/xmake/assets/47598093/aa01c56d-c47c-4263-9bdd-eb7dc6bceffd)

设置交叉编译：
```bash
$xmake f -p cross --sdk=/usr/arm-linux-gnueabi --arch=arm64 -vD
configure
{
    plat = cross
    kind = static
    sdk = /usr/arm-linux-gnueabi
    ccache = true
    target_os = linux
    mode = release
    buildir = build
    ndk_stdcxx = true
    arch = arm64
    host = linux
}
```

对应的测试 target：
```lua
target("cctest", function()
    set_kind("binary")
    add_files("test/cctest.c")
end)
```

编译命令没有出现 SDK 的任何内容：

![image](https://github.com/xmake-io/xmake/assets/47598093/2ef83b99-fd0d-4b2b-843c-a2f1bd480522)

理论上应该是 clang -target aarch64-linux-gnu --sysroot=/usr/arm-linux-gnueabi cctest.c -o cctest ？

**Comment by waruqi** [comment]
试下

```console
xmake f -p cross --cross=arm-linux-gnueabi- --sdk=/usr -c
xmake -rv
```

设置 -target 和 sysroot ，你只能自定义 toolchain

**Reply by TOMO-CAT** [reply]
在 toolchain 里设置对应的编译链接参数后能编译单个简单的 target，但是安装 xmake-repo 库的时候报错了：
```bash
xmake f -p cross --cross=aarch64-linux-gnu- --arch=armv8-a --toolchain=aarch-linux-gun --sdk=/usr -cvD
```
对应的 xmake.lua 文件：
```bash
toolchain("aarch-linux-gun", function()
    on_load(function(toolchain)
        os.setenv("C_INCLUDE_PATH", "")
        toolchain:add("cxflags",
                      "-m64 -fPIC --sysroot=/usr/aarch64-linux-gnu --target=aarch64-linux-gnu -march=armv8-a")
        toolchain:add("ldflags", "-m64 --target=aarch64-linux-gnu -march=armv8-a")
        toolchain:add("shflags",
                      "-m64 -fPIC --sysroot=/usr/aarch64-linux-gnu --target=aarch64-linux-gnu -march=armv8-a")
    end)
end)

target("binary", function()
    set_kind("binary")
    add_files("test/test.c")
end)
```
![image](https://github.com/xmake-io/xmake/assets/47598093/954c2c18-625f-4e4c-9595-62d56d9dffe6)

安装 protobuf 的时候出问题了，看着像是工具链本身的问题：
![image](https://github.com/xmake-io/xmake/assets/47598093/b793b678-8555-4917-836f-3a1d259024e8)

之前是把 ldflags 里的 sysroot 删掉了，但是现在还是不行，我琢磨下怎么做。

**Reply by TOMO-CAT** [reply]
https://bugs.archlinux.org/task/62394
![image](https://github.com/xmake-io/xmake/assets/47598093/6b2b1297-0d93-4d3d-a88d-5a2b187a1c09)
看样子是个 bug，得更换工具链或者手动修改 sdk 了。
